internal/client/client.go

```diff
@@ -478,41 +572,606 @@ func (c *Client[H, Hasher, N, E, Header]) Block(hash H) (*generic.SignedBlock[N,
+	aux api.AuxDataOperations,
+	forkchoice common.ForkChoiceStrategy,
+	importExisting bool,
+) (common.ImportResult, error) {
+	parentHash := importHeaders.Post().ParentHash()
+	status, err := c.backend.Blockchain().Status(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	parentStatus, err := c.backend.Blockchain().Status(parentHash)
+	if err != nil {
+		return nil, err
+	}
+
+	parentExists := parentStatus != blockchain.BlockStatusUnknown
+
+	if !importExisting && status == blockchain.BlockStatusInChain {
+		return common.ImportResultAlreadyInChain{}, nil
+	}
+
+	info := c.backend.Blockchain().Info()
+	gapBlock := info.BlockGap != nil && info.BlockGap.Start == importHeaders.Post().Number()
+
+	// the block is lower than our last finalized block so it must revert
+	// finality, refusing import.
+	if status == blockchain.BlockStatusUnknown && importHeaders.Post().Number() < info.FinalizedNumber && !gapBlock {
+		return nil, blockchain.ErrNotInFinalizedChain
+	}
+
+	// this is a fairly arbitrary choice of where to draw the line on making notifications,
+	// but the general goal is to only make notifications when we are already fully synced
+	// and get a new chain head.
+	makeNotifications := origin == consensus.NetworkBroadcastBlockOrigin ||
+		origin == consensus.OwnBlockOrigin ||
+		origin == consensus.ConsensusBroadcastBlockOrigin
+
+	var finalStorageChanges *api.StorageChanges
+
+	if storageChanges != nil {
+		switch sc := (*storageChanges).(type) {
+		case common.Changes[H, Hasher]:
+			err := c.backend.BeginStateOperation(operation.Op, parentHash)
+			if err != nil {
+				return nil, err
+			}
+
+			if c.config.OffchainIndexingApi {
+				err := operation.Op.UpdateOffchainStorage(sc.OffchainStorageChanges)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			err = operation.Op.UpdateDBStorage(sc.Transaction)
+			if err != nil {
+				return nil, err
+			}
+
+			err = operation.Op.UpdateStorage(sc.MainStorageChanges, sc.ChildStorageChanges)
+			if err != nil {
+				return nil, err
+			}
+
+			err = operation.Op.UpdateTransactionIndex(sc.TransactionIndexChanges)
+			if err != nil {
+				return nil, err
+			}
+
+			finalStorageChanges = &api.StorageChanges{
+				StorageCollection:      sc.MainStorageChanges,
+				ChildStorageCollection: sc.ChildStorageChanges,
+			}
+		case common.Import[H]:
+			strg := storage.Storage{}
+			for _, state := range sc.State {
+				if len(state.ParentStorageKeys) == 0 && len(state.StateRoot) == 0 {
+					for _, entry := range state.KeyValues {
+						strg.Top.Set(string(entry.StorageKey), entry.StorageValue)
+					}
+				} else {
+					for _, parentStorage := range state.ParentStorageKeys {
+						var storageKey []byte
+						prefixedStorageKey := storage.PrefixedStorageKey(parentStorage)
+						if childType := storage.NewChildTypeFromPrefixedKey(prefixedStorageKey); childType != nil {
+							storageKey = childType.Key
+						} else {
+							return nil, blockchain.ErrInvalidChildStorageKey
```

Comment:
Can you wrap in `blockchain.ErrBackend` instead?
```suggestion
							return nil, fmt.Errorf("%w: Invalid child storage key", blockchain.ErrBackend)
```